Home 화면 리팩토링

song1900 opened the pull request:
HomeViewController의 기능을 extension으로 분리시켰던 방법에서 UICollectionView을 관리하는 클래스를 생성해서 코드를 분리시켜봤습니다.
- CollectionView를 관리하는 업데이트하는 메서드와 프로퍼티를 HomeCollectionViewHandler 클래스로 옮겼습니다.
- HomeViewController의 크기는 줄어들고 이전보다 관리하기가 용이해졌습니다.
- Cell의 레이아웃을 설정하는 코드를 Cell로 옮겼습니다.

song1900 committed:
Refactor: NSCollectionLayoutItem을 생성하는 코드를 Cell로 이동

song1900 committed:
Refactor: HomeCollectionViewHandler 생성

HomeViewController에서 CollectionView를 관리하고 업데이트하는 메서드와 프로퍼티들을 HomeCollectionViewHandler class로 이동시켜 뷰컨트롤러를 깔끔하게 유지할 수 있도록 했습니다.

f-lab-jercy commented on GIL/GIL/Screen/Home/HomeCollectionViewHandler.swift:
```diff
@@ -0,0 +1,113 @@
+    }
+    private enum Item: Hashable {
+        case search
+        case recentSearchPlace([PlaceData])
```
여기서 placeData를 직접 들고 있는것은 별로 좋지 못한것 같습니다. 데이터 모델을 따로두고 사용하는 형태로 하면 좋을것 같습ㄴ디ㅏ.

song1900 replied:
https://github.com/f-lab-edu/gil-navi-ios/pull/23/commits/c74a8f2bbe4f9dd1fea811d239f7e770f676fd6a
해당 커밋으로 수정했습니다.

f-lab-jercy commented on GIL/GIL/Screen/Home/HomeCollectionViewHandler.swift:
```diff
@@ -0,0 +1,113 @@
+        for collectionView: UICollectionView,
+        at indexPath: IndexPath
+    ) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: HomeSearchCollectionViewCell.reuseIdentifier, for: indexPath) as? HomeSearchCollectionViewCell else { return UICollectionViewCell() }
```
셀생성을 실패했을때 에러처리에 대해서 한번더 고민을 해보면 좋을것 같습니다.

song1900 replied:
1. 대체 셀을 반환하여 에러 메시지를 표시하는 방법
```swift
guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: HomeSearchCollectionViewCell.reuseIdentifier, for: indexPath) as? HomeSearchCollectionViewCell else {
    let errorCell = collectionView.dequeueReusableCell(withReuseIdentifier: ErrorCollectionViewCell.reuseIdentifier, for: indexPath) as! ErrorCollectionViewCell
    errorCell.errorMessage = "Error.."
    return errorCell
}
``` 
2. HomeSearchCollectionViewCell을 생성하여 반환하는 방법
```swift
guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: HomeSearchCollectionViewCell.reuseIdentifier, for: indexPath) as? HomeSearchCollectionViewCell else {
    let fallbackCell = HomeSearchCollectionViewCell(frame: CGRect(x: 0, y: 0, width: collectionView.bounds.width, height: 50))
    fallbackCell.errorMessage = "Error.."
    return fallbackCell
}
``` 

이렇게 두 가지 방법을 생각해 봤습니다.

f-lab-jercy commented on GIL/GIL/Screen/Home/HomeCollectionViewHandler.swift:
```diff
@@ -0,0 +1,113 @@
+        let newItems = [Item.recentSearchPlace(data)]
+        snapshot.appendItems(newItems, toSection: .main)
+        dataSource.apply(snapshot, animatingDifferences: true)
+        resetLayout()
```
layout reset을 매번 하고 있는데 왜 매번 진행하게 작업이 되어 있을까요?

song1900 replied:
HomeViewController의 viewWillAppear 메서드가 호출될 때마다 SwiftData에서 검색 데이터를 불러오도록 되어있습니다.
검색 데이터는 0 ~ 3개까지 개수가 달라질 수 있고 개수에 따라 Cell의 높이가 변경되어야 해서 레이아웃을 다시 그리도록 해뒀습니다.

song1900 committed:
Merge branch 'main' of https://github.com/f-lab-edu/gil-navi-ios into feature/refactor-home

song1900 committed:
Merge branch 'feature/view-layout' into feature/refactor-home

song1900 committed:
Refactor: Item 열거형에서 [PlaceData]를 직접 포함하는 코드 수정

데이터를 관리하는 별도의 프로퍼티를 생성해서 데이터를 관리할 수 있는 방식으로 코드를 수정했습니다.

song1900 committed:
Chore: ConfigureCell 메서드 수정

기존 UICollectionViewCell에서 UICollectionViewCell?로 변경하여 옵셔널을 허용 하도록 수정했습니다.
언래핑 하는 코드도 삭제했습니다.

song1900 committed:
Merge branch 'feature/refactor-home' of https://github.com/f-lab-edu/gil-navi-ios into feature/refactor-home

song1900 committed:
Merge branch 'main' of https://github.com/f-lab-edu/gil-navi-ios into feature/refactor-home
